flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1264,6 +1273,483 @@ def test_all_transitions(self, desired_state, discovered_state):
+    :type local_applications: iterable of ``Application``s
+    :param additonal_node_config: Additional nodes to include in the cluster
+        configration.
+    :type additional_node_config: ``set`` of ``Node``s
+    :param Leases leases: Leases to include in the cluster configration.
+    """
+    calculator = RecordingCalculator(NoOp())
+    deployer = deployer.set(calculator=calculator)
+    cluster_configuration = Deployment(
+        nodes={
+            Node(
+                uuid=deployer.node_uuid,
+                hostname=deployer.hostname,
+                manifestations={
+                    manifestation.dataset.dataset_id: manifestation
+                    for manifestation in desired_manifestations
+                },
+            ),
+        } | additional_node_config,
+        leases=leases,
+    )
+
+    local_state = BlockDeviceDeployerLocalState(
+        node_uuid=deployer.node_uuid,
+        hostname=deployer.hostname,
+        datasets={dataset.dataset_id: dataset
+                  for dataset in local_datasets},
+    )
+    node_state, nonmanifest_datasets = local_state.shared_state_changes()
+    cluster_state = DeploymentState(
+        nodes={node_state.set(applications=local_applications)},
+        nonmanifest_datasets=nonmanifest_datasets.datasets,
+    )
+
+    deployer.calculate_changes(
+        configuration=cluster_configuration,
+        cluster_state=cluster_state,
+        local_state=local_state,
+    )
+    case.assertEqual(
+        {dataset.dataset_id: dataset
+         for dataset in expected_datasets},
+        calculator.desired_datasets,
+    )
+
+
+class CalculateDesiredStateTests(SynchronousTestCase):
+    """
+    Tests for ``BlockDeviceDeployer._calculate_desired_state``.
+    """
+    def setUp(self):
+        self.hostname = ScenarioMixin.NODE
+        self.node_uuid = ScenarioMixin.NODE_UUID
+        self.api = UnusableAPI()
+        self.deployer = BlockDeviceDeployer(
+            node_uuid=self.node_uuid,
+            hostname=self.hostname,
+            block_device_api=self.api,
+            mountroot=mountroot_for_test(self),
+        )
+
+    def test_no_manifestations(self):
+        """
+        If there are no Manifestations on this node, then
+        there are no desired datasets calculated.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[],
+            expected_datasets=[],
+        )
+
+    def test_manifestation(self):
+        """
+        If there is a manifesation on this node, then the
+        corresponding dataset has a desired state of ``MOUNTED``.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[ScenarioMixin.MANIFESTATION.transform(
+                ['dataset', 'metadata'], ScenarioMixin.METADATA,
+            )],
+            expected_datasets=[
+                DesiredDataset(
```

Comment:
This chunk seems like it's repeated a lot, perhaps we can have a default expected `DesiredDataset` which gets reused in the test bodies?